redgenes/utils.py

```diff
@@ -1,29 +1,138 @@
+
+    If the field does not exist, returns 'NA' or 0 depending on the specified
+    dtype; If the field has a list of values, returns the first item of the
+    list in the specified dtype. See https://github.com/hyattpd/prodigal/wiki/understanding-the-prodigal-output for an explaination of qualifier key values
+    """
+    if isinstance(tmp := qualiers_dict.get(qualifier_key), list):
+        # If qualifier is a list, take the first value
+        qualifier_value = tmp[0]
+    else:
+        qualifier_value = qualiers_dict.get(qualifier_key, None)
+
+    if qualifier_value:
+        # Returns the qualifier value in the correct dtype
+        if qualifier_value == "None":
+            return None
+        return qualifier_dtype(qualifier_value)
+    else:
+        return qualifier_value
+
+
+def process_qualifiers(qualifiers_dict, dtype_map):
+    """Extract selected qualifier fields from SeqRecord.feature.qualifiers in specified dtypes."""
+    processed_qualifers = []
+    for field, dtype in dtype_map.items():
+        value = extract_qualifier(qualifiers_dict, field, dtype)
+        processed_qualifers.append(value)
+    return processed_qualifers
+
+
+################################
+# Run bash commands
+################################
+def run_command(commands, errortype, errortext):
+    """Run bash commands using subprocess."""
     try:
-        res = subprocess.run(commands)
+        res = subprocess.run(commands, capture_output=True, check=True)
         assert res.returncode == 0
-        return
     except Exception as e:
-        raise (f"There is an {errortext}: {e}")
+        raise errortype(f"There is an {errortext}: {e}.")
 
 
+################################
+# Zip and unzip fasta files
+################################
 def run_unzip_fna(filepath):
-    commands = [
-        "gzip",
-        "-d",
-        f"{filepath}",
-    ]
-    run_command(commands, "error")
+    """Checks if a zipped file exists and unzip it."""
+    if Path(filepath).exists():
+        commands = [
+            "gzip",
+            "-d",
+            f"{filepath}",
+        ]
+        run_command(commands, InvalidFna, "error")
+    elif (unzipped_path := Path(filepath.replace(".gz", ""))).exists():
+        raise ValueError(f"Unzipped file already exists: {str(unzipped_path)}.")
+    else:
+        raise FileNotFoundError(f"File not found: {filepath}.")
 
 
 def run_zip_fna(filepath):
+    """Checks if an unzipped file exists and zip it."""
+    if not Path(filepath).exists():
+        raise FileNotFoundError(f"Unzipped file does not exist: {filepath}.")
+
+    zipped_path = filepath + ".gz"
+    if Path(zipped_path).exists():
+        raise ValueError(f"Zipped file already exists: {zipped_path}.")
+
     commands = [
         "gzip",
-        "-c",
         f"{filepath}",
     ]
-    run_command(commands, "error")
+    run_command(commands, InvalidFna, "error")
+
+
+@contextmanager
+def copy_and_unzip(zip_path, tmp_dir):
```

Comment:
This is really nice :) buuuuut, if the underlying programs just accept .gz then that would be better